### PR TITLE
Support the `template_from_string()` Twig function

### PR DESCRIPTION
**Pull request by richardhj**
## Issue description:

Contao's Twig extensions crash if one wants to use the `template_from_string()` function.

Details: https://twig.symfony.com/doc/3.x/functions/template_from_string.html

## How to reproduce:

**services.xml:**

```xml
<service id="Twig\Extension\StringLoaderExtension">
    <tag name="twig.extension"/>
</service>
```

**my_template.html.twig:**

```twig

{% set title = "{{ item.name }}" %}

<p>Title: {{ include(template_from_string(title)) }}</p>
```

## Encountered error:

`Object of class Twig\TemplateWrapper could not be converted to string`

The Twig extensions internally makes use of the TemplateWrapper to dynamically build templates.

## Proposed solution:

Contao's Twig extensions shall ignore such made-up templates.

### CODE REVIEW

File: core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php [hunk 1]
```diff
@@ -54,9 +54,15 @@ public function getTag(): string
...
      * Return the adjusted logical name or the unchanged input if it does not
      * match the Contao Twig namespace.
+     *
+     * TODO change signature to public static function adjustTemplateName(string|TemplateWrapper $name, TemplateHierarchyInterface $hierarchy): string|TemplateWrapper in Contao 5.1
      */
-    public static function adjustTemplateName(string $name, TemplateHierarchyInterface $hierarchy): string
+    public static function adjustTemplateName($name, TemplateHierarchyInterface $hierarchy)
     {
+        if (!is_string($name)) {
+            return $name;
```
Comment (ausi) [inline core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php]:
```suggestion
            if ($name instanceof TemplateWrapper) {
                return $name;
            }

            throw new \TypeError(sprintf('Argument 1 passed to %s() must be of the type string|TemplateWrapper, %s given', __METHOD__, get_debug_type($name)));
```

File: core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php [hunk 2]
```diff
@@ -54,9 +54,15 @@ public function getTag(): string
     /**
      * Return the adjusted logical name or the unchanged input if it does not
      * match the Contao Twig namespace.
+     *
+     * TODO change signature to public static function adjustTemplateName(string|TemplateWrapper $name, TemplateHierarchyInterface $hierarchy): string|TemplateWrapper in Contao 5.1
```
Comment (ausi) [inline core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php:59]:
```suggestion
     * TODO change signature to public static function adjustTemplateName(string|TemplateWrapper $name, TemplateHierarchyInterface $hierarchy): string|TemplateWrapper in Contao 5.1
     *
     * @param string|TemplateWrapper $name
     *
     * @return string|TemplateWrapper
```

**Review by ausi — COMMENTED** [review]
I think we should correctly check the types in Contao 4.13.
Apart from that, approved ✅

**Comment by m-vo** [comment]
Apart from the fix, we should maybe also point out to the Twig devs, that the variable of the original `include` function is currently annotated with `string|array` which seems to be wrong then: https://github.com/twigphp/Twig/blob/58f7b2208151d41259f9b42751779172a1235513/src/Extension/CoreExtension.php#L1307

But maybe we need to cater for the array case as well  (if applicable) :thinking:.

File: core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php [hunk 3]
```diff
@@ -54,9 +54,23 @@ public function getTag(): string
...
+     *
+     * @param string|TemplateWrapper $name
```
Comment (stof) [inline core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php:61]: looks like you are missing a use statement for TemplateWrapper

**Comment by richardhj** [comment]
RTM :)

File: core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php [hunk 4]
```diff
@@ -54,9 +55,23 @@ public function getTag(): string
     /**
      * Return the adjusted logical name or the unchanged input if it does not
      * match the Contao Twig namespace.
+     *
+     * TODO change signature to public static function adjustTemplateName(string|TemplateWrapper $name, TemplateHierarchyInterface $hierarchy): string|TemplateWrapper in Contao 5.1
```
Comment (Toflar) [inline core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php:59]: 5.1?

Comment (ausi) [reply core-bundle/src/Twig/Inheritance/DynamicIncludeTokenParser.php:59]: This TODO meant for the upstream merge I think.

**Review by ausi — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @richardhj.

**Pull request merged**